### DISCUSSION TITLE
[@types/chrome] It seems like in reality chrome.scripting.ExecutionWorld is an enum that can contain both ISOLATED and MAIN

### DISCUSSION THREAD

**Discussion by danielhjacobs** [DefinitelyTyped/DefinitelyTyped#69776] [Issues with a @types package]
Currently https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/chrome/index.d.ts#L7992 defines it as a string with either the value "MAIN" or "ISOLATED", but in reality Chrome and Firefox Nightly (https://bugzilla.mozilla.org/show_bug.cgi?id=1736575) have both `chrome.scripting.ExecutionWorld.MAIN` and `chrome.scripting.ExecutionWorld.ISOLATED` defined and Firefox stable has just `chrome.scripting.ExecutionWorld.ISOLATED` defined.

I'm not sure, but this might also apply to https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/webextension-polyfill/namespaces/scripting.d.ts#L157 and https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/firefox-webext-browser/index.d.ts#L4736

https://developer.chrome.com/docs/extensions/reference/api/scripting#type-ExecutionWorld says that it's an ENUM.

It should be more like this:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/chrome/index.d.ts#L12179-L12193

[a comment by an automated account is not shown]

**Comment by danielhjacobs** [comment]
/** The JavaScript world for a script to execute within. */
    export enum ExecutionWorld {
        ISOLATED = "ISOLATED",
        MAIN = "MAIN",
    }

**Reply by danielhjacobs** [reply]
See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69778

**Reply by danielhjacobs** [reply]
My PR was closed without merge.

Will I permanently need to keep the hack at https://github.com/ruffle-rs/ruffle/blob/e1efe90dbd42b42b975557d572327c7c0aa6d89a/web/packages/extension/src/utils.ts#L22-L29, https://github.com/ruffle-rs/ruffle/blob/e1efe90dbd42b42b975557d572327c7c0aa6d89a/web/packages/extension/src/utils.ts#L107, and https://github.com/ruffle-rs/ruffle/blob/e1efe90dbd42b42b975557d572327c7c0aa6d89a/web/packages/extension/src/utils.ts#L149 so TS won't complain about the check at https://github.com/ruffle-rs/ruffle/blob/e1efe90dbd42b42b975557d572327c7c0aa6d89a/web/packages/extension/src/background.ts#L14 ?

Since FF just added support for ExecutionWorld MAIN in FF beta I can imagine a lot of people adding such a check to not break their Firefox extensions on older versions of Firefox by only adding the scripts on browsers with MAIN ExecutionWorld support.

That check can't currently be added without hacking around the incorrect DefinitelyTyped type definitions.

**Comment by matcarlson** [comment]
You could try making the types like:

```ts
  interface RegisteredContentScript {
     // ...
    world?: ExecutionWorld | `${ExecutionWorld}`;
  }
```

This should allow chrome.scripting.StyleOrigin.AUTHOR or "AUTHOR" to pass the tests, and shouldn't break the tests. The one caveat is that we may need to bump the TypeScript version to 4.1. To keep it completely backwards-compatible you'd need to use `ExecutionWorld | 'MAIN' | 'ISOLATED';`, a minor inconvenience IMO. 

FYI, most DefinitelyTyped types lag behind the software we try to document as it's maintained mostly by volunteers with functionality added by users of the library. The maintainers just try to keep the library as universally-usable as possible through code reviews and some overall strategy.
